Spatial index functionality

OliviaYtterbrink opened the pull request:
Builds on #10766 and needs #10780 

changelog: Index-backed support for spatial point equality and range searches from Cypher.

For example, the following query will return all points in the specified range:

    WITH point({latitude: 56.5, longitude: 12.5}) as bottomLeft, point({latitude: 57.5, longitude: 13.5}) as topRight
    MATCH (c:City) WHERE c.location > bottomLeft AND c.location < topRight

This also includes support for storing points and arrays of points as node properties.

OliviaYtterbrink committed:
Expand fusion index support to include spatial index for points

OliviaYtterbrink committed:
Dynamic handling of of different spatial indexes

Handle indexing of points with different coordinate
reference system. Only creating files for actually used
indexes.

craigtaverner committed:
Make parts of Fusion Index available to Spatial for dynamic fusion

craigtaverner committed:
Abstract out common behaviour for dynamic creation in KnownSpatialIndex

This class should represent all access to an on-disk GBPTree for a spatial index specific to a particular CoordinateReferenceSystem. The SpatialFusion* classes will multiplex to the relevant index based on CRS, and the SpatialSchemaIndexProvider will maintain a common collection of all KnownSpatialIndex instances for use in the multiplexing logic.

craigtaverner committed:
Working spatial index populating and surviving restarts

OliviaYtterbrink committed:
Apply filter on index lookups for geometry

Spatial index needs post processing because it will have potential false
positive hits due to multi dimensional to one dimensional transformation

OliviaYtterbrink committed:
Added test

OliviaYtterbrink committed:
Fix compareTo for points close to each other

OliviaYtterbrink committed:
Use correct comparison for points to enable rangeSeek

OliviaYtterbrink committed:
Extract common functionality

Native Number and Spatial indexes share a lot of functionality.

OliviaYtterbrink committed:
Change handling of max and min points

OliviaYtterbrink committed:
Fixed handling of different type ranges

Also fixed miss on close and error message

craigtaverner committed:
Cleaned up geometry range queries to use correct CRS

craigtaverner committed:
More careful test of inequality index seeks for spatial index

craigtaverner committed:
Make sure min/max range based on same CRS

craigtaverner committed:
Integrate hilbert space filling curve into spatial index

craigtaverner committed:
Add SpatialIndex tests

OliviaYtterbrink committed:
Fix errors after rebase

OliviaYtterbrink committed:
Updated license header to 2018

OliviaYtterbrink committed:
WIP added some notes about what is failing

OliviaYtterbrink committed:
Accept spatial in newapi

OliviaYtterbrink committed:
Removed `type` from spatial key

This was not really needed, since we already know what CRS the index is
using when reading/writing.

craigtaverner committed:
WIP - Re-integrated hilbert with multi-range search

craigtaverner committed:
WIP - Use cursor filter for post-filtering range queries for spatial

craigtaverner committed:
WIP - revert post-index filtering and disable tests for this

OliviaYtterbrink committed:
Use correct limit

OliviaYtterbrink committed:
Move test that is not supported in spatial

OliviaYtterbrink committed:
Fix values for spatial testing

OliviaYtterbrink committed:
Increased maxlevel for curve

This value has to be benchmarked, but 5 was too low granularity
for the cartesian points, would return (100000,-100000) when
searching for points less than (0,0) due to it ending up in a
tile close to the center line.

OliviaYtterbrink committed:
Removed debug prints

OliviaYtterbrink committed:
Added spatial fusion tests

Lojjs commented on community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala:
```diff
@@ -223,12 +223,14 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
-    val optNumericRange = groupedRanges.get(classOf[Number]).map(_.asInstanceOf[InequalitySeekRange[Number]])
-    val optStringRange = groupedRanges.get(classOf[String]).map(_.mapBounds(_.toString))
+    val optNumericRange: Option[InequalitySeekRange[Number]] = groupedRanges.get(classOf[Number]).map(_.asInstanceOf[InequalitySeekRange[Number]])
+    val optStringRange: Option[InequalitySeekRange[String]] = groupedRanges.get(classOf[String]).map(_.mapBounds(_.toString))
```
Unnecessary type annotation (two rows)

Lojjs commented on community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala:
```diff
@@ -332,6 +338,33 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
+        }
+      }
+  }).getOrElse(Iterator.empty)
+
```
IndexSeekByGeometryRange and IndexSeekByStringRange have a lot of duplicate code. Have you considered to break out some parts to a common subfunction?

OliviaYtterbrink replied:
It would end up harder to read.

Lojjs commented on community/values/src/main/java/org/neo4j/values/storable/Values.java:
```diff
@@ -301,6 +308,21 @@ public static PointArray pointArray( Point[] points )
+            }
+            values[i] = Values.point( (PointValue) maybePoint );
+        }
+        return new PointArray.Direct( values );
```
Seems to be the same as `return pointArray(values)`, maybe use that instead?

OliviaYtterbrink committed:
Checkstyle

Lojjs commented on community/values/src/main/java/org/neo4j/values/storable/PointValue.java:
```diff
@@ -154,7 +154,7 @@ else if ( this.coordinate.length < other.coordinate.length )
         for ( int i = 0; i < coordinate.length; i++ )
         {
-            int cmpVal = (int) (this.coordinate[i] - other.coordinate[i]);
+            int cmpVal = Double.compare(this.coordinate[i], other.coordinate[i]);
```
Nice find!

Lojjs commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/state/TxState.java:
```diff
@@ -1044,6 +1046,81 @@ public PrimitiveLongReadableDiffSets indexUpdatesForRangeSeekByNumber( IndexDesc
     }
 
+    @Override
+    public PrimitiveLongReadableDiffSets indexUpdatesForRangeSeekByGeometry( IndexDescriptor descriptor,
```
A lot of duplicate code in indexUpdatesForRangeSeekByGeometry and indexUpdatesForRangeSeekByNumber. Consider to break out some of the code to a common sub-function.

Lojjs commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/LayoutTestUtil.java:
```diff
@@ -140,7 +142,17 @@ private Value existingNonUniqueValue( RandomRule randomRule )
     }
 
-    private IndexEntryUpdate<IndexDescriptor>[] generateAddUpdatesFor( Number[] values )
+    IndexEntryUpdate<IndexDescriptor>[] someSpatialUpdatesNoDuplicateValues()
```
This function is never used, can be removed.

Lojjs commented on community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/IndexQuery.java:
```diff
@@ -243,6 +263,7 @@ public IndexQueryType type()
         @Override
         public boolean acceptsValue( Value value )
         {
+            // TODO Handle pointvalue here?
```
Make sure that this is included in a Spatial Trello card if you still plan to do it (otherwise remove the TODO or fix it if it is easy)

OliviaYtterbrink replied:
Already carded.

Lojjs commented on community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/IndexQuery.java:
```diff
@@ -347,6 +368,88 @@ public boolean toInclusive()
+            else if ( value instanceof LongValue )
+            {
+                // This is called from the hilbert space filling curve index which decomposed 2D range into 1D long curve index
+                // TODO: Filter with property lookup
```
Make sure that this is included in a Spatial Trello card if you still plan to do it (otherwise remove the TODO or fix it if it is easy)

Lojjs commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NativeHitIndexProgressor.java:
```diff
@@ -52,6 +52,7 @@ public boolean next()
                 KEY key = seeker.get().key();
                 if ( client.acceptNode( key.getEntityId(), key.asValue() ) )
                 {
+                    // TODO this part fails for SpatialSchemaKey.asValue()
```
Maybe you should fix this?

OliviaYtterbrink replied:
This is tied to the other todo about exact query, which is carded.

Lojjs commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/SpatialIndexAcceptanceTest.scala:
```diff
@@ -0,0 +1,173 @@
+    testPointRead("MATCH (p:Place) WHERE p.location = point({latitude: 56.7, longitude: 12.78, crs: 'WGS-84'}) RETURN p.location as point", Values.pointValue(CoordinateReferenceSystem.WGS84, 12.78, 56.7))
+  }
+
+  private def testPointRead(query: String, expected: PointValue*): Unit = {
```
`testPointRead` and `testPointScan` are identical apart from the word "NodeIndexSeek"/"NodeIndexScan". I would merge them into one function `testPoint` with a flag/boolean for seek/scan

Lojjs reviewed (commented):
Overall, very good work! Because of the size of the PR  I haven't deep dived on all things (I leave that for @burqen ) but I have some comments

Lojjs commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/SpatialIndexAcceptanceTest.scala:
```diff
@@ -0,0 +1,173 @@
+
+import scala.collection.Map
+
+class SpatialIndexAcceptanceTest extends CypherFunSuite with GraphDatabaseTestSupport {
```
You have changed all `createLabeledNode` followed by `MATCH ... SET ... RETURN` to `CREATE ... SET ... RETURN` in this file. I'm thinking that there may be a value of checking that setting points works both with `CREATE` and `MATCH` so maybe some tests should use one approach and some the other.

OliviaYtterbrink committed:
Limit testing on deep levels due to performance

OliviaYtterbrink committed:
Fixups from review comments

OliviaYtterbrink committed:
Extract common code for range seek

OliviaYtterbrink committed:
Clarified todo comments

sherfert committed:
Equality and inequality post-filtering for spatial index seeks.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/PropertyValueComparison.java:
```diff
@@ -52,6 +53,8 @@ private PropertyValueComparison()
 
     public static final PropertyValueComparator<Object> COMPARE_STRINGS = new StringPropertyValueComparator();
 
+    public static final PropertyValueComparator<PointValue> COMPARE_POINTS = new PointPropertyValueComparator();
```
It looks like we are missing an update on `AnyPropertyValueComparator` and `SuperType`. None of them consider geometric. Please update `PropertyValueComparison` test accordingly.

OliviaYtterbrink replied:
Fixed

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java:
```diff
@@ -994,6 +1000,25 @@ private PrimitiveLongResourceIterator filterIndexStateChangesForRangeSeekByNumbe
+            return nodes.augmentWithRemovals( labelPropertyChangesForGeometry.augment( nodeIds ) );
+        }
+        return nodeIds;
+
```
Empty line

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexUtils.java:
```diff
@@ -0,0 +1,76 @@
+     * @param <E> the type of exception anticipated, inferred from the lambda
+     * @throws E
+     */
+    public static <E extends Exception> void forAll( ActionAble actionable, Object... accessors ) throws E
```
This is really nice! Although parameter name `accessors` is a bit misleading. Maybe `subjects` or `targets`?

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexUtils.java:
```diff
@@ -0,0 +1,76 @@
+     * @param accessors varargs array of objects to call the function on
+     * @param <E> the type of exception anticipated, inferred from the lambda
+     * @throws E
+     */
```
Sweet ass javadoc!

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/SpatialFusionIndexUpdater.java:
```diff
@@ -0,0 +1,145 @@
+        this.accessor = null;
+    }
+
+    SpatialFusionIndexUpdater( Map<CoordinateReferenceSystem,SpatialKnownIndex> indexMap, long indexId, SpatialKnownIndex.Factory indexFactory,
```
I would prefer if both of those constructors could delegate to common private constructor. That way it's more easier to confirm what initial state is and also easier to maintain.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/SpatialFusionIndexAccessor.java:
```diff
@@ -0,0 +1,177 @@
+        {
+            try
+            {
+                indexReaders.put( index.getKey(), index.getValue().getOnlineAccessor(descriptor, samplingConfig).newReader() );
```
The fact that we create our spatial fusion readers here with a static set of readers is worrying. Here is why.

1. Create fusion reader, `fr`
2. Updater add value for a new `CoordinateReferenceSystem`.
3. `fr.query` towards this new CRS that it will not see because it has no corresponding reader.
4. No profit :(

We are especially vulnerable against this because `fr` will be cached in `KernelStatement` and live for the whole transaction. We could use a similar pattern like what updater and populator does. "Given my set of readers and this predicate, please select the correct reader for me or update me with a missing reader."

Please add test for this scenario :D

burqen replied:
What are the plans for this fellow?

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/SpatialFusionIndexAccessor.java:
```diff
@@ -0,0 +1,177 @@
+        {
+            try
+            {
+                allEntriesReader.add( index.getOnlineAccessor( descriptor, samplingConfig ).newAllEntriesReader() );
```
The same is actually not true for this reader, because it is only used by consistency checker with no concurrent updates. Please add javadoc to interface if you decide to utilise this assumption.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/SpatialFusionIndexAccessor.java:
```diff
@@ -0,0 +1,177 @@
+            @Override
+            public Iterator<Long> iterator()
+            {
+                return new CombiningIterable( allEntriesReader ).iterator();
```
Unchecked assignment

Lojjs replied:
We have tried to rewrite as a checked assignment, but we can't figure out how that should be done without turning the warning into an error

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/SpatialFusionIndexPopulator.java:
```diff
@@ -0,0 +1,143 @@
+    {
+        // When an index is first created using `CREATE INDEX` or `graph.schema.for*` then a call will be made here,
+        // however, the KnownSpatialIndex will not yet exist, since that can only be created on demand
+        assert indexMap.isEmpty();
```
I don't know if you are aware, but assertions are only "turned on" for testing. In production this will be a `no-op`. If you consider this check to be mucho importante then you need to do a real `throw`. Right now I have a hard time deciding what would be the "right thing". What do  you think?

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/SpatialFusionIndexPopulator.java:
```diff
@@ -0,0 +1,143 @@
+        for ( Map.Entry<CoordinateReferenceSystem,SpatialKnownIndex> index : indexMap.entrySet() )
+        {
+            batchMap.put( index.getKey(), new ArrayList<>() );
+        }
```
`selectUpdates` below will `computeIfAbsent` so we don't need to pre populate this map with lists here.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/SpatialFusionIndexReader.java:
```diff
@@ -0,0 +1,158 @@
+            for ( IndexReader reader : readerMap.values() )
+            {
+                iterators[i] = reader.query( predicates[0] );
+                ++i;
```
This is fine but I would prefer `iterators[i++]` since that is a much more common style in the code base 😄

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/SpatialFusionIndexUpdater.java:
```diff
@@ -0,0 +1,145 @@
+        {
+            return updater;
+        }
+        if ( mode != null )
```
This implicit dependency would be nice to have documented. To clarify it even more I suggest using static constructor methods for this guy to clarify the different use cases.
 
Something like this maybe?
```
    static SpatialFusionIndexUpdater updaterForPopulator( Map<CoordinateReferenceSystem,SpatialKnownIndex> indexMap, long indexId, SpatialKnownIndex.Factory indexFactory,
            IndexDescriptor descriptor, IndexSamplingConfig samplingConfig, PropertyAccessor accessor )
    {
        return new SpatialFusionIndexUpdater( indexMap, indexId, indexFactory, descriptor, samplingConfig, accessor, null );
    }

    static SpatialFusionIndexUpdater updaterForAccessor( Map<CoordinateReferenceSystem,SpatialKnownIndex> indexMap, long indexId, SpatialKnownIndex.Factory indexFactory,
            IndexDescriptor descriptor, IndexSamplingConfig samplingConfig, IndexUpdateMode mode )
    {
        return new SpatialFusionIndexUpdater( indexMap, indexId, indexFactory, descriptor, samplingConfig, null, mode );
    }

    private SpatialFusionIndexUpdater( Map<CoordinateReferenceSystem,SpatialKnownIndex> indexMap, long indexId, SpatialKnownIndex.Factory indexFactory,
            IndexDescriptor descriptor, IndexSamplingConfig samplingConfig, PropertyAccessor accessor, IndexUpdateMode mode )
    {
        this.indexMap = indexMap;
        this.indexId = indexId;
        this.indexFactory = indexFactory;
        this.descriptor = descriptor;
        this.samplingConfig = samplingConfig;
        this.mode = mode;
        this.accessor = accessor;
    }
```

OliviaYtterbrink replied:
✅

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/SpatialFusionSchemaIndexProvider.java:
```diff
@@ -0,0 +1,239 @@
+
+    private Map<Long,Map<CoordinateReferenceSystem,SpatialKnownIndex>> indexes = new HashMap<>();
+
+    public SpatialFusionSchemaIndexProvider( PageCache pageCache, FileSystemAbstraction fs, IndexDirectoryStructure.Factory directoryStructure, Monitor monitor,
```
Line break

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/SpatialFusionSchemaIndexProvider.java:
```diff
@@ -0,0 +1,239 @@
+                k -> new SpatialKnownIndex( directoryStructure(), crs, indexId, pageCache, fs, monitor, recoveryCleanupWorkCollector ) );
+    }
+
+    private void findAndCreateKnownSpatialIndexes()
```
From my understanding of how spatial fusion is supposed to work I understand what this guy tries to do. It would be nice to have some comments in the code though. Like expected path we are currently on for each layer or something like that. Does this make sense?
Building up the path as we go along like:
```
// somerootdirectory/
// somerootdirectory/andsomedirectoryindicatingid/
// somerootdirectory/andsomedirectoryindicatingid//andthenfinalyspecificindex
```
Also, more specifically. What will the directory structure look like?

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/SpatialFusionSchemaIndexProvider.java:
```diff
@@ -0,0 +1,239 @@
+                                if ( !index.indexExists() )
+                                {
+                                    // TODO: log something here?
+                                    indexesFor( indexId ).remove( crs );
```
Is it really safe to just ignore a potentially missing index here? Does this not mean that you can remove a whole subdirectory corresponding to some CRS and it will still be happy about this and not report a failed state?

craigtaverner replied:
We were thinking there are only two states, the index exists, or it doesn't. Are you suggesting we try to deal with cases where, for some reason, a directory has been made (or an index deleted, but leaving directories)?

I had assumed a FAILED state was associated with an index that actually did exist, but the contents were broken for some reason (out of disk-space or power failure during populating perhaps). The case where some directories exist, bt not the index itself, appears to me to be a non-existing index.

burqen replied:
This is how I think it works, please correct me if I'm wrong here 🗡 
1. We only create a directory for a given CRS when we actually insert some value there.
2. We never delete index files, even if they are empty.
Conclusion, if there is a directory there must be an index file. And if there is not, there is no way for us to determine if that file had some data that is now missing or not.
We expect data to be their, it is not, thus I think this would be considered as FAILED.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialSchemaIndexReader.java:
```diff
@@ -0,0 +1,162 @@
+        throw new UnsupportedOperationException( "Cannot initialize 1D range in multidimensional spatial index reader" );
+    }
+
+    public void query( IndexProgressor.NodeValueClient cursor, IndexOrder indexOrder, IndexQuery... predicates )
```
Missing `@Override`

sherfert committed:
Fix ClassCastExceptions

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialKnownIndex.java:
```diff
@@ -0,0 +1,216 @@
+    private SpaceFillingCurve curve;
+
+    /**
+     * Create a representation of a spatial index for a specific coordinate reference system. This constructure should be used for first time creation.
```
This ~constructure~ constructor and also long line.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialKnownIndex.java:
```diff
@@ -0,0 +1,216 @@
+        // Depends on crs
+        SchemaIndexProvider.Descriptor crsDescriptor =
+                new SchemaIndexProvider.Descriptor( Integer.toString( crs.getTable().getTableId() ), Integer.toString( crs.getCode() ) );
+        IndexDirectoryStructure indexDir = IndexDirectoryStructure.directoriesBySubProvider( directoryStructure ).forProvider( crsDescriptor );
```
Long line

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialKnownIndex.java:
```diff
@@ -0,0 +1,216 @@
+        curve = new HilbertSpaceFillingCurve2D( envelopeFromCRS( crs ), 8 );
+    }
+
+    static Envelope envelopeFromCRS( CoordinateReferenceSystem crs )
```
Quite a lot of `todo` here. I suppose those are for some later PR?

OliviaYtterbrink replied:
Yes, they are slated for later work.

craigtaverner replied:
Proper management of coordinate reference systems and in particular the extents required by the space filling curves are planned in upcoming work.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialKnownIndex.java:
```diff
@@ -0,0 +1,216 @@
+
+    private static String indexFileName( long indexId )
+    {
+        return "index-" + indexId;
```
If one logical index with id 1 has multiple different sub-indexes on different CRS. The files will all be called `index-1` but live under different directories named after the CRS. Is this correct?
If so, directory structure will be something like
```
whatever-provider-name/1/native/index-1
whatever-provider-name/1/lucene/...bunch-of-lucene-files
whatever-provider-name/1/spatial/whatever-CRS/index-1
whatever-provider-name/1/spatial/other-CRS/index-1
whatever-provider-name/1/spatial/third-CRS/index-1
```
Does it make more or less sense to remove the intermediate layer of directories for spatial and use crs-name as file name?
```
whatever-provider-name/1/native/index-1
whatever-provider-name/1/lucene/...bunch-of-lucene-files
whatever-provider-name/1/spatial/whatever-CRS-1
whatever-provider-name/1/spatial/other-CRS-1
whatever-provider-name/1/spatial/third-CRS-1
```

craigtaverner replied:
We are planning to also save metadata in a second file in each subdirectory together with the GBPTree, and then the advantage of having the index in its own directory is increased. Metadata will contain, for example, the extents of the index. This allows for future features like automatic re-indexing of inefficiently created indexes, or at least Cypher planner support for potentially inefficient indexes.

burqen replied:
That makes sense! ✅

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialKnownIndex.java:
```diff
@@ -0,0 +1,216 @@
+        return fs.fileExists( indexFile );
+    }
+
+    public String readPopupationFailure() throws IOException
```
~readPopupationFailure~ readPopulationFailure

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialKnownIndex.java:
```diff
@@ -0,0 +1,216 @@
+
+    public void closePopulator( IndexDescriptor descriptor, IndexSamplingConfig samplingConfig, boolean populationCompletedSuccessfully ) throws IOException
+    {
+        getPopulator( descriptor, samplingConfig ).close( populationCompletedSuccessfully );
```
This is a good example of why `getPopulator` is risky. If we call `close` twice we end up deleting our index.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialKnownIndex.java:
```diff
@@ -0,0 +1,216 @@
+        }
+    }
+
+    public IndexPopulator getPopulator( IndexDescriptor descriptor, IndexSamplingConfig samplingConfig )
```
This method is in general super risky because it has the side effect of deleting the current index if the field happens to be `null`. I would prefer if we could decouple the intentions of getting the assumed existing populator from lazily creating it.

craigtaverner replied:
Similar `indexExists()` checks to what we do in getAccessor could deal with the two cases we should care about.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialKnownIndex.java:
```diff
@@ -0,0 +1,216 @@
+    private final SchemaIndexProvider.Monitor monitor;
+    private final RecoveryCleanupWorkCollector recoveryCleanupWorkCollector;
+    private SpatialSchemaIndexAccessor indexAccessor;
+    private NativeSchemaIndexPopulator indexPopulator;
```
Over all comment about this class:
It is not super clear what the lifecycle looks like for this guy, especially around `indexAccessor` and `indexPopulator`. It looks like they live side by side and that has some problems:

1. Populator will have a writer open against the underlying GBPTree. So will updater created by accessor have. This is not possible since the GBPTree only support a single writer.
2. The logical lifecycle for an index is to transfer from populating state to either online or failure. The populator and accessor is logically bound to separate states and thus should not overlap.

Please reach out if you want help to clear this up 😄

craigtaverner replied:
I would think the state management code higher up the stack would prevent this, but we should probably add a little defensive code here too. This is slightly related to the other comments about the two ways we use the populator (one for background population, and one for initializing a new index that is needed only for the accessor).

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialLayout.java:
```diff
@@ -0,0 +1,105 @@
+/**
+ * {@link Layout} for numbers where numbers doesn't need to be unique.
+ */
+public abstract class SpatialLayout extends Layout.Adapter<SpatialSchemaKey,NativeSchemaValue>
```
Copy-paste javadoc 🙈

burqen commented on community/spatial-index/src/main/java/org/neo4j/gis/spatial/index/curves/SpaceFillingCurve.java:
```diff
@@ -189,6 +206,14 @@ private Long derivedValueFor( double[] coord, int level )
+    /**
+     * Given a derived key, find the normalized coordinate it corresponds to on a specific level
+     */
+    public long[] normalizedCoordinateFor( long derivedValue )
```
Method is never used

craigtaverner replied:
I checked and this method has actually been deleted in the original code in 'Neo4j Spatial', so it appeared to have been used only temporarily.

burqen commented on community/spatial-index/src/main/java/org/neo4j/gis/spatial/index/curves/SpaceFillingCurve.java:
```diff
@@ -143,6 +143,23 @@ private Long derivedValueFor( double[] coord, int level )
         assertValidLevel( level );
         long[] normalizedValues = getNormalizedCoord( coord );
+        return derivedValueFor( normalizedValues, level );
+    }
```
We create quite a lot of objects in this class. I think some of them could be converted to reused fields instead to save some object allocation and garbage collection. In general the schema index package and index module go through a lot of effort to keep object creation on the low.

I would like to see this being addressed moving forward, not necessarily in this pr though.

craigtaverner replied:
Worth thinking about. But the current code should be multi-thread safe and the suggested change would break that, which is a disadvantage also worth thinking about.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialLayout.java:
```diff
@@ -0,0 +1,105 @@
+
+    @Override
+    public SpatialSchemaKey copyKey( SpatialSchemaKey key,
+            SpatialSchemaKey into )
```
Could live on single line

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialLayoutNonUnique.java:
```diff
@@ -0,0 +1,62 @@
+
+public class SpatialLayoutNonUnique extends SpatialLayout
+{
+    private static final String IDENTIFIER_NAME = "NUNI";
```
Please find a unique identifier for spatial layouts. Number use `"NUNI"` and `"UNI"` for non unique and unique respectively.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialSchemaKey.java:
```diff
@@ -0,0 +1,181 @@
+     * for comparison within the space filling index as long as the original spatial range has already
+     * been decomposed into a collection of 1D curve ranges before calling down into the GPTree.
+     */
+    int compareValueTo( SpatialSchemaKey other )
```
Sweet javadoc!

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialLayoutUnique.java:
```diff
@@ -0,0 +1,73 @@
+ */
+public class SpatialLayoutUnique extends SpatialLayout
+{
+    private static final String IDENTIFIER_NAME = "UNI";
```
Find new unique identifier.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialLayoutUnique.java:
```diff
@@ -0,0 +1,73 @@
+    public static final int MINOR_VERSION = 1;
+    public static long IDENTIFIER = Layout.namedIdentifier( IDENTIFIER_NAME, NativeSchemaValue.SIZE );
+
+    public SpatialLayoutUnique( CoordinateReferenceSystem crs, SpaceFillingCurve curve )
```
Can be package-private

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialLayoutNonUnique.java:
```diff
@@ -0,0 +1,62 @@
+    public static final int MINOR_VERSION = 1;
+    public static long IDENTIFIER = Layout.namedIdentifier( IDENTIFIER_NAME, NativeSchemaValue.SIZE );
+
+    public SpatialLayoutNonUnique( CoordinateReferenceSystem crs, SpaceFillingCurve curve )
```
Can be package-private

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialSchemaIndexReader.java:
```diff
@@ -0,0 +1,162 @@
+                KEY treeKeyFrom = layout.newKey();
+                KEY treeKeyTo = layout.newKey();
+                treeKeyFrom.fromDerivedValue( Long.MIN_VALUE, range.min );
+                treeKeyTo.fromDerivedValue( Long.MAX_VALUE, range.max + 1 );
```
Are we sure `range.max` can never be `Long.MAX_VALUE`?

OliviaYtterbrink replied:
Yes, but we can add a test that shows this

Lojjs replied:
Test added ✅

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialSchemaKey.java:
```diff
@@ -0,0 +1,181 @@
+/**
+ * Includes value and entity id (to be able to handle non-unique values).
+ * A value can be any {@link PointValue} and is represented as a {@code long} to store the 1D mapped version
+ * and a type describing the geomtry type and number of dimensions.
```
~geomtry~ geometry

Also, javadoc describe a type that I can't find in the code. Has this changed?

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialSchemaKey.java:
```diff
@@ -0,0 +1,181 @@
+        return (PointValue) values[0];
+    }
+
+    private void extractRawBitsAndType( PointValue value )
```
Update method, not writing type.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialSchemaKey.java:
```diff
@@ -0,0 +1,181 @@
+    /**
+     * Extracts raw bits and type from a {@link PointValue} and store as state of this {@link SpatialSchemaKey} instance.
+     */
+    private void writePoint( CoordinateReferenceSystem crs, double[] coordinate )
```
Javadoc mention non existing type.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/state/TxState.java:
```diff
@@ -1031,17 +1033,58 @@ public PrimitiveLongReadableDiffSets indexUpdatesForRangeSeekByNumber( IndexDesc
-                                      selectedUpper, selectedIncludeUpper ).values();
-        for ( PrimitiveLongDiffSets diffForSpecificValue : inRange )
+    @Override
+    public PrimitiveLongReadableDiffSets indexUpdatesForRangeSeekByGeometry( IndexDescriptor descriptor,
```
All other variants of this method extract the necessary fields from predicate before passing them in. Can we follow that same pattern here?

craigtaverner replied:
Or alternatively, how about changing the other methods to take predicates? I had the impression this was slightly cleaner code, and the reason we did not do this elsewhere was historical, but I could be wrong.

burqen replied:
There is no one happier than me if you want to clean that up 😍

burqen replied:
Although there seem to be quite a lot of tests that also would need to be updated. And thinking about it, testability might have been the reason for not using the predicates to begin with. Only guessing ofc.

Lojjs replied:
Me and Olivia ended up following the same pattern as you @burqen

burqen commented on community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/IndexQuery.java:
```diff
@@ -347,6 +367,88 @@ public boolean toInclusive()
+            {
+                // This is called from the hilbert space filling curve index which decomposed 2D range into 1D long curve index
+                // This will return false positives that need to be filtered later
+                return true;
```
Dat comment 😻

burqen commented on community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/NativeLuceneFusionSchemaIndexProviderFactory.java:
```diff
@@ -87,12 +88,14 @@ public static FusionSchemaIndexProvider newInstance( PageCache pageCache, File s
         int priority = useNativeIndex ? PRIORITY : 0;
         return new FusionSchemaIndexProvider( nativeProvider,
-                luceneProvider, new NativeSelector(), DESCRIPTOR, priority, directoriesByProvider( storeDir ), fs );
+                spatialProvider, luceneProvider, new NativeSelector(), DESCRIPTOR, priority, directoriesByProvider( storeDir ), fs );
```
Adding spatial to current fusion index is potentially a format change and we need to think about what to do with that. Does not need to be a part of this PR though.

craigtaverner replied:
We should think this through a few times more, but based on our last discussion this does not seem to be a problem. Since the new types and new indexes are being added together, there does not exist a released version that could create a problem. For example, if we run in 3.4, add points to the database (and they are added to lucene-native and stored in the spatial index), then downgrade to 3.3 and add more points, they will not be stored in either the property store or the index (since 3.3 does not support that at all), so the spatial index remains hidden. Upgrading to 3.4 again will bring it back in a consistent state.

The same is obviously not true for a native string index, since the same type will be stored in lucene in older versions, so you have to be explicit about the type of index.

burqen replied:
I think you might be right about this! Let's re-iterate later.

burqen commented on community/spatial-index/src/main/java/org/neo4j/gis/spatial/index/curves/SpaceFillingCurve.java:
```diff
@@ -143,6 +143,23 @@ private Long derivedValueFor( double[] coord, int level )
+    /**
+     * Given a normalized coordinate in multiple dimensions, calculate its derived key for maxLevel
+     */
+    public Long derivedValueFor( long[] normalizedValues )
```
Can be package-private

craigtaverner replied:
This code comes from `Neo4j Spatial` where the value is placed into an explicit lucene index, not the GPBTree, and this method is required there. Moving forward instead of depending on Neo4j Spatial code, it will depend on Neo4j code, and this public API is part of the contract.

burqen replied:
Aah, I see. Can we document this to remind us not to "clean" this up and break `Neo4j Spatial` in the future?

Lojjs replied:
Added comment on the method and the top of the file ✅

burqen commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/LayoutTestUtil.java:
```diff
@@ -140,7 +96,12 @@ private Value existingNonUniqueValue( RandomRule randomRule )
     }
 
-    private IndexEntryUpdate<IndexDescriptor>[] generateAddUpdatesFor( Number[] values )
+    IndexEntryUpdate<IndexDescriptor>[] someSpatialUpdatesWithDuplicateValues()
```
Since this class now handle spatial values to we should differentiate when we only talk about numbers as well. For example, `ALL_EXTREME_VALUES` only handle number values and could be renamed.

Lojjs replied:
Seems out-dated, this code looks different now

burqen commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialLayoutTestUtil.java:
```diff
@@ -0,0 +1,139 @@
+    Value asValue( Number value )
+    {
+        // multiply value with 5 to make sure the points aren't too close together, avoiding false positives that will be filtered in higher level later
+        return Values.pointValue( CoordinateReferenceSystem.WGS84, 3 * value.doubleValue(), 3 * value.doubleValue() );
```
Those are mighty fine 5s ;)

burqen reviewed (changes requested):
Alright, first walk through finished :D Exiting stuff!

Overall I think this is really good! It fits well into the current fusion index system and you follow the layed out patterns well which makes it easy to follow what is happening. Great!

I found what I think is some pretty severe concurrency issues that need to be addressed and (as expected for such a large PR) some places that could use some polishing. See individual comments.

Other thoughts:
1. IndexCapability. What are the plans for this moving forward?

2. You've been pretty thorough with testing around Fusion and and SpatialFusion. Awesome!

3. I didn't see any tests for the lazy creation of new spatial sub-indexes. This would be useful both full stack and on FusionSpatial level.

4. What I think is lacking the most right now is full stack integration testing. What are the plan for this moving forward? From the top of my mind, those are the things we need to test specifically with spatial type properties:
- Reads and writes through old kernel API
- Reads and writes through new kernel API
- Concurrency tests with reads and writes on both APIs
- Recovery
- Backup
- Consistency checker
- Integration in HA
- Integration in CC

5. We also need coverage for spatial in
- Soaks
- Ha-robustness
- CC-robustness
- Recovery-robustness

OliviaYtterbrink commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/LookupFilter.java:
```diff
@@ -59,18 +59,18 @@ public static PrimitiveLongIterator exactIndexMatches( PropertyAccessor accessor
         }
 
-        IndexQuery[] numericPredicates =
+        IndexQuery[] filteredPredicates =
```
We can't do any filtering for the geometric case here, since we don't have access to the actual value. Filtering is done on the cypher level (https://github.com/neo4j/neo4j/pull/10876/commits/8de5b4efb398332d2000beabe287ea5ce9902b61)

OliviaYtterbrink committed:
Minor fixes, typos etc from review

OliviaYtterbrink committed:
Add PointValue to ordering

Placed between boolean and Numbers

OliviaYtterbrink committed:
Added test that shows behaviour of range

Range.max should never end up being Long.MaxValue

OliviaYtterbrink committed:
change to private constructor

Makes it easier to reason about the usage.

OliviaYtterbrink committed:
Minor cleanup

OliviaYtterbrink committed:
Save query to be able to answer numberOfProperties

OliviaYtterbrink committed:
Fix filtering for spatial in kernel instead of cypher

also rename method

OliviaYtterbrink committed:
Fixed rebase issues

OliviaYtterbrink commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/LookupFilter.java:
```diff
@@ -59,18 +59,18 @@ public static PrimitiveLongIterator exactIndexMatches( PropertyAccessor accessor
         }
 
-        IndexQuery[] numericPredicates =
+        IndexQuery[] filteredPredicates =
```
Changed to filtering on kernel level. (https://github.com/neo4j/neo4j/pull/10702/commits/b89c467cb16bebaf3c0e70a433a1c12f8a883b4f)

OliviaYtterbrink committed:
Removed dead code, changed to throw exception

OliviaYtterbrink committed:
Mark index as failed if missing file

OliviaYtterbrink committed:
Change concept of SpatialKnownIndex

SpatialKnownIndex will no longer hand out accessor/populator,
instead it will be called for read/write/etc directly from
SpatialFusionIndexPopulator/Accessor.

OliviaYtterbrink committed:
clear the map after drop, since no more interaction should come

Lojjs committed:
Add comments about dependency from Neo4j Spatial

craigtaverner commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialKnownIndex.java:
```diff
@@ -91,6 +114,302 @@ public SpatialKnownIndex( IndexDirectoryStructure directoryStructure, Coordinate
+        treeKey = layout.newKey();
+        treeValue = layout.newValue();
+        schemaIndex = new NativeSchemaIndex<>( pageCache, fs, indexFile, layout, monitor, descriptor, indexId );
+        if ( uniqueSampler( descriptor ) )
```
Method should be called `isUnique`, not `uniqueSampler`.

craigtaverner commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialKnownIndex.java:
```diff
@@ -147,77 +466,37 @@ public InternalIndexState readState( IndexDescriptor descriptor ) throws IOExcep
     }
 
-    private NativeSchemaIndexPopulator makeIndexPopulator( IndexDescriptor descriptor, IndexSamplingConfig samplingConfig )
+    private boolean uniqueSampler( IndexDescriptor descriptor )
```
Method should be called `isUnique`, not `uniqueSampler`.

craigtaverner commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/SpatialFusionIndexAccessor.java:
```diff
@@ -53,43 +53,58 @@
     public void refresh() throws IOException
     {
-        forAll( entry -> ((SpatialKnownIndex) entry).getOnlineAccessor( descriptor, samplingConfig ).refresh(), indexMap.values().toArray() );
+        // not required in this implementation
```
Perhaps a comment about why?

craigtaverner commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/SpatialFusionIndexAccessor.java:
```diff
@@ -98,14 +113,8 @@ public IndexReader newReader()
-                e.printStackTrace();
-            }
+            // TODO should this be populated here, or delegate to SpatialFusionIndexReader?
+            indexReaders.put( index.getKey(), index.getValue().newReader( samplingConfig, descriptor ) );
```
Elsewhere we check the mode and complete initialization based on that. Why not here?

craigtaverner committed:
Increase test coverage for spatial index

One failing test ignored, wil be debugged.

burqen commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialLayoutTestUtil.java:
```diff
@@ -0,0 +1,139 @@
+import org.neo4j.values.storable.Value;
+import org.neo4j.values.storable.Values;
+
+class SpatialLayoutTestUtil extends LayoutTestUtil<SpatialSchemaKey,NativeSchemaValue>
```
This guy is not used anymore. What happened to all the tests?

Lojjs replied:
The file that it tested was removed when we rewrote SpatialKnownIndex. This was used for testing the Spatial Index Accessor which is now happening on the fusion level. So I guess the right thing is just to remove this unused test util.

burqen replied:
✅

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/SpatialFusionSchemaIndexProvider.java:
```diff
@@ -215,8 +217,7 @@ private void findAndCreateKnownSpatialIndexes()
                                 {
-                                    // TODO: log something here?
-                                    indexesFor( indexId ).remove( crs );
+                                    index.markAsFailed( "Index file was not found" );
```
Nice

Lojjs committed:
Move state logic into SpatialKnownIndex

Lojjs committed:
Review comments and clean up.

Lojjs committed:
Add unit tests for SpatialKnownIndex

burqen commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialKnownIndexTest.java:
```diff
@@ -0,0 +1,215 @@
+        storeDir = new File( "store" );
+        fs.deleteRecursively( storeDir );
+        fs.mkdir( storeDir );
+        indexDir = new File( new File( new File( new File( new File( storeDir, "schema" ), "index" ), "spatial-1.0" ), "1" ), "1-4326" );
```
`indexDir = FileUtils.path( "schema", "index", "spatial-1.0", "1", "1-4326" );`
What's up with the name "1-4326"?

Lojjs replied:
The file is named after the CRS official naming which consists of 2 tables with an id for each CRS. 4326 is the id of WGS84 (longitude/latitude) which is used in this test

burqen commented on community/kernel/src/test/java/org/neo4j/kernel/api/index/NodeUpdatesTest.java:
```diff
@@ -78,6 +83,7 @@ public void shouldNotGenerateUpdateForMultipleExistingPropertiesAndLabels()
         NodeUpdates updates = NodeUpdates.forNode( nodeId, labels )
                 .existing( propertyKeyId1, Values.of( "Neo" ) )
                 .existing( propertyKeyId2, Values.of( 100L ) )
+                .existing( propertyKeyId3, Values.of( Values.pointValue( CoordinateReferenceSystem.WGS84, 12.3, 45.6 ) ) )
```
Don't need the extra `Values.of` wrap.

burqen commented on community/kernel/src/test/java/org/neo4j/kernel/api/index/NodeUpdatesTest.java:
```diff
@@ -116,15 +122,17 @@ public void shouldGenerateUpdatesForLabelAdditionWithExistingProperties()
                 NodeUpdates.forNode( nodeId, empty, labels )
                         .existing( propertyKeyId1, Values.of( "Neo" ) )
                         .existing( propertyKeyId2, Values.of( 100L ) )
+                        .existing( propertyKeyId3, Values.of( Values.pointValue( CoordinateReferenceSystem.WGS84, 12.3, 45.6 ) ) )
```
Don't need the extra `Values.of` wrap.

burqen commented on community/kernel/src/test/java/org/neo4j/kernel/api/index/SimpleIndexAccessorCompatibility.java:
```diff
@@ -27,6 +27,9 @@
 import org.neo4j.kernel.api.schema.index.IndexDescriptorFactory;
+import org.neo4j.values.storable.CoordinateReferenceSystem;
+import org.neo4j.values.storable.PointValue;
+import org.neo4j.values.storable.Values;
```
Don't need those imports.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialKnownIndex.java:
```diff
@@ -0,0 +1,556 @@
+
+    // POPULATING
+
+    public synchronized void close( boolean populationCompletedSuccessfully ) throws IOException
```
Easy to confuse with `close()`. Maybe call this method `finishPopulation` or something similar.

burqen commented on community/kernel/src/test/java/org/neo4j/graphdb/IndexingAcceptanceTest.java:
```diff
@@ -353,6 +359,10 @@ public void shouldBeAbleToQuerySupportedPropertyTypes()
         assertCanCreateAndFind( db, LABEL1, property, 12L );
         assertCanCreateAndFind( db, LABEL1, property, (float)12. );
         assertCanCreateAndFind( db, LABEL1, property, 12. );
+        assertCanCreateAndFind( db, LABEL1, property, new SpatialMocks.MockPoint( 12.3, 45.6, mockWGS84() ) );
```
`new SpatialMocks.MockPoint(...)` could be exchanged for a static method instead `SpatialMocks.mockPoint(...)` and let constructor be private.
Also constructor of `SpatialMocks.MockGeometry` can then be private.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialKnownIndex.java:
```diff
@@ -0,0 +1,556 @@
+
+    public interface Factory
+    {
+        SpatialKnownIndex selectAndCreate( Map<CoordinateReferenceSystem,SpatialKnownIndex> indexMap, long indexId, Value value );
```
This method is just a shortcut to `selectAndCreate( indexMap, indexId, crs )` and it's only used in one place in one test. I think it's better to just get rid of it. This will also simplify mocking.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/SpatialFusionSchemaIndexProvider.java:
```diff
@@ -0,0 +1,229 @@
+    public SpatialKnownIndex selectAndCreate( Map<CoordinateReferenceSystem,SpatialKnownIndex> indexMap, long indexId, CoordinateReferenceSystem crs )
+    {
+        return indexMap.computeIfAbsent( crs,
+                k -> new SpatialKnownIndex( directoryStructure(), crs, indexId, pageCache, fs, monitor, recoveryCleanupWorkCollector ) );
```
Use the key provided to the mapping function instead of directly referencing the method parameter.
`crsKey -> new SpatialKnownIndex( directoryStructure(), crsKey, indexId, pageCache, fs, monitor, recoveryCleanupWorkCollector ) );`

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialKnownIndex.java:
```diff
@@ -0,0 +1,556 @@
+        schemaIndex.assertOpen();
+        try
+        {
+            return singleUpdater.initialize( schemaIndex.tree.writer(), true );
```
I feel confused about the fact that we create a new `schemaIndex.tree.writer()` here while we also have a `singleTreeWriter` field.

OliviaYtterbrink replied:
the singleTreeWriter is alive while populating and is closed when calling `finishPopulation`, while the writer that is returned with the newUpdater should be closed from the `IndexUpdater.close`.

It might be that we should tell the updater that it should not handle the closing of the writer and instead we reuse the `singleTreeWriter` but I think that's for a later cleanup.

burqen replied:
Yep, later cleanup is fine by me.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialKnownIndex.java:
```diff
@@ -0,0 +1,556 @@
+
+    // POPULATING
+
+    public synchronized void close( boolean populationCompletedSuccessfully ) throws IOException
```
Why do we need to synchronize on `close`, `drop` and `create`? Can we document this?

OliviaYtterbrink replied:
Discussed live, drop and create can be called at the same time.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialKnownIndex.java:
```diff
@@ -0,0 +1,556 @@
+
+    private void assertPopulatorOpen()
+    {
+        if ( schemaIndex.tree == null )
```
This does not verify that populator is open.

burqen reviewed (changes requested):
Found a few more irregularities.

OliviaYtterbrink committed:
Cleaned up methods

burqen commented:
@OliviaYtterbrink @Lojjs Those last changes hit the nail! Let's just wait for the tests to go through before merging 👍

Lojjs committed:
Inequality fix for 3.4

Was there before but removed due to human error
